lib/obj.js

```diff
@@ -1,365 +1,479 @@
+    addNode();
+
+    // Parse the obj file
+    return readLines(objPath, parseLine)
+        .then(function() {
+            // Unload resources
+            positions = undefined;
+            normals = undefined;
+            uvs = undefined;
+
+            // Load materials and images
+            return finishLoading(nodes, mtlPaths, objPath, options);
+        });
+}
+
+function finishLoading(nodes, mtlPaths, objPath, options) {
+    nodes = cleanNodes(nodes);
+    if (nodes.length === 0) {
+        throw new RuntimeError(objPath + ' does not have any geometry data');
+    }
+    return loadMaterials(mtlPaths, objPath, options)
+        .then(function(materials) {
+            var imagePaths = getImagePaths(materials);
+            return loadImages(imagePaths, objPath, options)
+                .then(function(images) {
+                    return {
+                        nodes : nodes,
+                        materials : materials,
+                        images : images
+                    };
+                });
         });
+}
+
+function getAbsolutePath(mtlPath, objPath) {
+    if (!path.isAbsolute(mtlPath)) {
+        mtlPath = path.join(path.dirname(objPath), mtlPath);
+    }
+    return mtlPath;
+}
+
+function outsideDirectory(filePath, objPath) {
+    return (path.relative(path.dirname(objPath), filePath).indexOf('..') === 0);
+}
 
-        stream.on('end', function () {
-            resolve({
-                vertexCount: vertexCount,
-                vertexArray: vertexArray,
-                positionMin: positionMin,
-                positionMax: positionMax,
-                hasUVs: hasUVs,
-                hasNormals: hasNormals,
-                materialGroups: materialGroups,
-                materials: materials,
-                images: images
+function loadMaterials(mtlPaths, objPath, options) {
+    var materials = {};
+    return Promise.map(mtlPaths, function(mtlPath) {
+        mtlPath = getAbsolutePath(mtlPath, objPath);
+        if (options.secure && outsideDirectory(mtlPath, objPath)) {
+            options.logger('Could not read mtl file at ' + mtlPath + ' because it is outside of the obj directory and the secure flag is true. Using default material instead.');
+            return;
+        }
+        return loadMtl(mtlPath)
+            .then(function(materialsInMtl) {
+                materials = combine(materials, materialsInMtl);
+            })
+            .catch(function() {
+                options.logger('Could not read mtl file at ' + mtlPath + '. Using default material instead.');
             });
-        });
+    }).then(function() {
+        return materials;
     });
 }
 
-function getImages(inputPath, materials) {
-    // Collect all the image files from the materials
-    var images = [];
+function loadImages(imagePaths, objPath, options) {
+    var images = {};
+    return Promise.map(imagePaths, function(imagePath) {
```

Comment:
I assume `loadImage` reads an image from disk.  Anywhere we are using `Promise.map` that is using an external resources (like file handles in this case) we need to set a concurrency on `Promise.map` otherwise it's going to try and open too many files at once when imagePaths contains a large array of images.  Another option would be to use `mapSeries` but that would only handle one iteration at a time and probably be a little slower.  Even something small like 10 or 100 is fine.  We just need to avoid out of control asynchronicity.